Mono.Nat/Pmp/Searchers/PmpSearcher.cs

```diff
@@ -109,24 +111,28 @@ internal static PmpSearcher Create ()
 
         }
 
-        protected override async Task SearchAsync (IPAddress gatewayAddress, TimeSpan? repeatInterval, CancellationToken token)
+        protected override async void SearchAsync (IPAddress gatewayAddress, TimeSpan? repeatInterval, CancellationToken token)
         {
-            if (token == CancellationToken.None) {
-                token = Cancellation.Token;
-            }
-
             do {
-                var currentSearch = CancellationTokenSource.CreateLinkedTokenSource (token);
-                Interlocked.Exchange (ref CurrentSearchCancellation, currentSearch)?.Cancel ();
+                CurrentSearchCancellation?.Cancel ();
+                var currentSearch = CurrentSearchCancellation = CancellationTokenSource.CreateLinkedTokenSource (token);
 
                 try {
                     await SearchOnce (gatewayAddress, currentSearch.Token).ConfigureAwait (false);
                 } catch (OperationCanceledException) {
-                    token.ThrowIfCancellationRequested ();
+                    if (token.IsCancellationRequested)
+                        break;
                 }
+
                 if (!repeatInterval.HasValue)
                     break;
-                await Task.Delay (repeatInterval.Value, token).ConfigureAwait (false);
+
+                try {
+                    await Task.Delay (repeatInterval.Value, token).ConfigureAwait (false);
+                } catch (OperationCanceledException) {
+                    break;
+                }
+
             } while (true);
```

Comment:
```suggestion
            } while (!token.IsCancellationRequested);
```

Reply:
Sure why not! In practice it's one fewer exceptions thrown when cancellation occurs.